axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderLineTreeCalculationsServiceImpl.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Axelor Business Solutions
+ *
+ * Copyright (C) 2005-2023 Axelor (<http://axelor.com>).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.sale.service.saleorder;
+
+import com.axelor.apps.base.db.repo.ProductRepository;
+import com.axelor.apps.sale.db.SaleOrderLine;
+import com.axelor.apps.sale.db.SaleOrderLineTree;
+import com.axelor.apps.sale.db.repo.SaleOrderLineRepository;
+import com.axelor.apps.sale.db.repo.SaleOrderLineTreeRepository;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.util.List;
+import org.apache.commons.collections.CollectionUtils;
+
+public class SaleOrderLineTreeCalculationsServiceImpl
+    implements SaleOrderLineTreeCalculationsService {
+  protected SaleOrderLineTreeRepository saleOrderLineTreeRepository;
+  protected ProductRepository productRepository;
+
+  protected SaleOrderLineRepository saleOrderLineRepository;
+  BigDecimal price = new BigDecimal(0);
+  BigDecimal subTotalCostPrice = new BigDecimal(0);
```

Comment:
I don't think these should be defined as instance fields, why don't you simply make them local variables?

Also, it's better to use `BigDecimal.ZERO` instead of `new BigDecimal(0)`